pandas/_libs/tslibs/timedeltas.pyx

```diff
@@ -234,29 +234,35 @@ def array_to_timedelta64(object[:] values, unit='ns', errors='raise'):
     n = values.shape[0]
     result = np.empty(n, dtype='m8[ns]')
     iresult = result.view('i8')
+    found_string = False
 
     # Usually, we have all strings. If so, we hit the fast path.
     # If this path fails, we try conversion a different way, and
     # this is where all of the error handling will take place.
     try:
         for i in range(n):
+            if isinstance(values[i], str):
+                found_string = True
             if values[i] is NaT:
                 # we allow this check in the fast-path because NaT is a C-object
                 #  so this is an inexpensive check
                 iresult[i] = NPY_NAT
             else:
                 result[i] = parse_timedelta_string(values[i])
     except (TypeError, ValueError):
-        unit = parse_timedelta_unit(unit)
+        parsed_unit = parse_timedelta_unit(unit or 'ns')
         for i in range(n):
             try:
-                result[i] = convert_to_timedelta64(values[i], unit)
+                result[i] = convert_to_timedelta64(values[i], parsed_unit)
             except ValueError:
                 if errors == 'coerce':
                     result[i] = NPY_NAT
                 else:
                     raise
 
+    if found_string and unit is not None:
```

Comment:
actually i don't think you even need to wait until the end, can you just do this in the same check on L251, eg.

```
if values[i] isNat:
   ....
else:
   if istance(values[i], str) and unit is None:
      raise.......
   result[i] = parse_....
```

Reply:
is there a test that actually hits this as well?

Reply:
It wouldn't work raising inside the `try` block with `ValueError` being in `except`

Reply:
The added test hits this when I checked with the repl

Reply:
I noticed that this is insufficient because we might overlook elements due to exception in the loop.
Actually it's cleaner to have a separate for-loop outside for the check, without affecting performance